Wikipedia/Code/WMFZeroConfigurationManager.m

```diff
@@ -0,0 +1,166 @@
+NSString *const WMFZeroOnDialogShownOnce = @"ZeroOnDialogShownOnce";
+NSString *const WMFZeroWarnWhenLeaving = @"ZeroWarnWhenLeaving";
+NSString *const WMFZeroXCarrier = @"X-Carrier";
+NSString *const WMFZeroXCarrierMeta = @"X-Carrier-Meta";
+
+@interface WMFZeroConfigurationManager ()
+
+@property (nonatomic, strong, readonly) WMFZeroConfigurationFetcher *zeroConfigurationFetcher;
+@property (nonatomic, strong, nullable, readwrite) WMFZeroConfiguration *zeroConfiguration;
+
+@property (atomic, copy, nullable) NSString* previousPartnerXCarrier;
+@property (atomic, copy, nullable) NSString* previousPartnerXCarrierMeta;
+@property (atomic, readwrite) BOOL isZeroRated;
+
+@end
+
+@implementation WMFZeroConfigurationManager
+@synthesize isZeroRated = _isZeroRated;
+@synthesize zeroConfigurationFetcher = _zeroConfigurationFetcher;
+
++ (void)load {
+    [super load];
+    //    FBTweakAction(@"Networking", @"Wikipedia Zero", @"Reset ZeroOnDialogShownOnce", ^{
+    //        [[NSUserDefaults wmf_userDefaults] setBool:NO forKey:WMFZeroOnDialogShownOnce];
+    //        [[NSUserDefaults wmf_userDefaults] synchronize];
+    //    });
+}
+
+- (WMFZeroConfigurationFetcher *)zeroConfigurationFetcher {
+    if (!_zeroConfigurationFetcher) {
+        _zeroConfigurationFetcher = [[WMFZeroConfigurationFetcher alloc] init];
+    }
+    return _zeroConfigurationFetcher;
+}
+
+- (void)setIsZeroRated:(BOOL)isZeroRated {
+    @synchronized(self) {
+        if(_isZeroRated != isZeroRated){
+            _isZeroRated = isZeroRated;
+            [[NSNotificationCenter defaultCenter] postNotificationName:WMFZeroRatingChanged object:self];
+        }
+    }
+}
+
+- (BOOL)isZeroRated {
+    BOOL isZeroRated;
+    @synchronized(self) {
+        isZeroRated = _isZeroRated;
+    }
+    return isZeroRated;
+}
+
+- (void)setWarnWhenLeaving:(BOOL)warnWhenLeaving {
+    [[NSUserDefaults wmf_userDefaults] setObject:[NSNumber numberWithBool:warnWhenLeaving]
+                                          forKey:WMFZeroWarnWhenLeaving];
+    [[NSUserDefaults wmf_userDefaults] synchronize];
+}
+
+- (BOOL)warnWhenLeaving {
+    return [[NSUserDefaults wmf_userDefaults] boolForKey:WMFZeroWarnWhenLeaving];
+}
+
+#pragma mark - Banner Updates
+
+/**
+ *   This method:
+ *
+ * - Fetches carrier specific strings placing them in self.zeroConfiguration object.
+ *
+ * - If the fetched zeroConfiguration has a nil "message" string, this means even though
+ *   there was a header, leading us to believe this network was Zero rated, its Zero
+ *   rating is not presently enabled. (It would be nice if the query fetching the
+ *   zeroConfiguration returned an "enabled" key/value, but it doesn't - it nils out the
+ *   values instead apparently.) So in the nil message case we set isZeroRated "NO".
+ */
+- (void)fetchZeroConfigurationAndSetIsZeroRatedIfNecessary {
+    
+    // Note: don't nil out self.zeroConfiguration in this method
```

Comment:
Instead of this comment, why not remove the `nullable` attribute from the zeroConfiguration property? If we don't have one on launch since it needs to be downloaded from the network, you can make yourself a "default" one. Try to favor using language constructs to enforce behavior over comments.

This will be nice to hold some of the default logic you have if a the configuration is nil like: WMFAppViewController line 884